stdlib/tests/collections/smt.rs

```diff
@@ -0,0 +1,162 @@
+    let expected = [ZERO; WORD_SIZE];
+    assert_smt_get_opens_correctly(key, expected, initial_root, store, &advice_map);
+
+    // check included leaves for all tiers
+    for depth in [16, 32, 48, 64] {
+        // compute the index and node value
+        let index = key[3].as_int();
+        let index = index >> (64 - depth);
+        let index = NodeIndex::new(depth as u8, index).unwrap();
+        let remaining = get_smt_remaining_key(key, depth as u8);
+        let node =
+            Rpo256::merge_in_domain(&[remaining.into(), value.into()], Felt::new(depth)).into();
+
+        // setup the store and run the test
+        let mut store = MerkleStore::new();
+        let root = store.set_node(initial_root, index, node).unwrap().root;
+        let advice_map = [(node.into_bytes(), value.to_vec())];
+        assert_smt_get_opens_correctly(key, value, root, store, &advice_map);
+    }
+}
+
+#[test]
+fn smtget_opens_correctly_from_tree_with_sibling() {
+    // setup the test
+    let empty = EmptySubtreeRoots::empty_hashes(64);
+    let initial_root = Word::from(empty[0]);
+    let mut seed = 1 << 40;
+    let key = seeded_word(&mut seed);
+    let value = seeded_word(&mut seed);
+
+    // check empty tree
+    let store = MerkleStore::new();
+    let advice_map = [];
+    let expected = [ZERO; WORD_SIZE];
+    assert_smt_get_opens_correctly(key, expected, initial_root, store, &advice_map);
+
+    // check included leaves for all tiers
+    for depth in [16, 32, 48, 64] {
+        // init a new storage
+        let mut store = MerkleStore::new();
+
+        // compute the index and node value
+        let index = key[3].as_int();
+        let index = index >> (64 - depth);
+        let index = NodeIndex::new(depth as u8, index).unwrap();
+        let remaining = get_smt_remaining_key(key, depth as u8);
+        let node =
+            Rpo256::merge_in_domain(&[remaining.into(), value.into()], Felt::new(depth)).into();
+
+        // insert the target node into the storage
+        let root = store.set_node(initial_root, index, node).unwrap().root;
+
+        // compute the index and sibling value
+        let sibling = key[3].as_int();
+        let sibling = sibling >> (64 - depth);
+        let sibling = sibling ^ 1;
+        let sibling = NodeIndex::new(depth as u8, sibling).unwrap();
+        let mut sibling_node = Word::from(node);
+        sibling_node[3] = sibling_node[3] + ONE;
+        let mut sibling_value = Word::from(value);
+        sibling_value[3] = sibling_value[3] + ONE;
+
+        // insert the sibling node into the storage
+        let root = store.set_node(root, sibling, sibling_node).unwrap().root;
+
+        // execute the test and expect the right node to be returned
+        let advice_map = [
+            (node.into_bytes(), value.to_vec()),
+            (sibling_node.into_bytes(), sibling_value.to_vec()),
+        ];
+        assert_smt_get_opens_correctly(key, value, root, store, &advice_map);
+    }
```

Comment:
I found this test somewhat difficult to follow. I think it would be simpler if we moved building of the `MerkleStore` out of the test and did it once. Then, we could test various lookup calls against this store. I'm even thinking that we should add a method to `MerkleStore` to build a tiered SMT from key-value pairs (but probably in a future PR).

The store could contain a single TSMT with 5 leaves (we could assign letters `a` though `e` to them - i.e. `a_key`, `a_value`). The leaves would be built as follows:
-`a` - no common key prefix with any other leaf. It would be located at depth 16.
- `b` and `c` - common key prefix for the first 16 bits (could go up to the first 31 bits). These leaves would be located at depth 32.
- `d` and `e` - common key prefix for the first 32 bits (could go up to the first 47 bits). These leaves would be located at depth 48.

We can't really test leaves at depth 64 yet as we don't have a mechanism to store two leaves which have the same 64-bit prefix (I'm actually not sure how this current test is working for depth 64). Or I guess we could instantiate such a tree, but the tests would currently fail.

We could even hardcode these keys, if it simplifies things. For example, the most-significant element of the keys could look like:
```Rust
let a_3 = 0b00000000_00000000_11111111_11111111_11111111_11111111_11111111_11111111;
let b_3 = 0b10000000_00000000_01111111_11111111_11111111_11111111_11111111_11111111;
let c_3 = 0b10000000_00000000_11111111_11111111_11111111_11111111_11111111_11111111;
let d_3 = 0b11000000_00000000_00000000_00000000_01111111_11111111_11111111_11111111;
let e_3 = 0b11000000_00000000_00000000_00000000_11111111_11111111_11111111_11111111;
```

Reply:
I added two more cases; `f` and `g`, that conflicts until the last tier

Reply:
Assuming the way described below is the correct way to compute the node value for the ordered list, then it is correct as the only unimplemented part would be if we have multiple conflicting leaves on the same node of depth `64`.

Mid-tier node value: $H_d(K, V)$
Bottom-tier node value: $H_64(K_0, V_0, K_1, V_1, K_2, V_2, (...), K_n, V_n)$

It means we have $H_64(K, V)$ for a single key/value pair that exists on a bottom-tier. Given it is not different than a mid-tier node value computation (i.e. $H_d(K, V)$), then it is valid.

Reply:
Regarding the `MerkleStore` initialization, I'm not sure I get it. The test case is the `MerkleStore` initialization; the only thing besides that is 

```rust
    // assert all nodes are returned with the MASM implementation
    let advice_map = [
        (a_n.into_bytes(), a_v.to_vec()),
        (b_n.into_bytes(), b_v.to_vec()),
        (c_n.into_bytes(), c_v.to_vec()),
        (d_n.into_bytes(), d_v.to_vec()),
        (e_n.into_bytes(), e_v.to_vec()),
        (f_n.into_bytes(), f_v.to_vec()),
        (g_n.into_bytes(), g_v.to_vec()),
    ];
    assert_smt_get_opens_correctly(a, a_v, root, store.clone(), &advice_map);
    assert_smt_get_opens_correctly(b, b_v, root, store.clone(), &advice_map);
    assert_smt_get_opens_correctly(c, c_v, root, store.clone(), &advice_map);
    assert_smt_get_opens_correctly(d, d_v, root, store.clone(), &advice_map);
    assert_smt_get_opens_correctly(e, e_v, root, store.clone(), &advice_map);
    assert_smt_get_opens_correctly(f, f_v, root, store.clone(), &advice_map);
    assert_smt_get_opens_correctly(g, g_v, root, store.clone(), &advice_map);
```

The creation of the keys can be somewhat automated as they are just a bunch of XORs, but I understood the goal was to make everything hardcoded and explicit as possible

Reply:
> Bottom-tier node value: $H_{64}(K_0, V_0, K_1, V_1, K_2, V_2, (...), K_n, V_n)$

We may end up doing it this way, but we may also do something different. I'd rather handle this when we work on the bottom tier. In my mind, our handling of the bottom tier is incomplete and I don't think we need tests for it (and ideally, anything that touches the bottom tier should panic).

> Regarding the `MerkleStore` initialization, I'm not sure I get it. The test case is the `MerkleStore` initialization; the only thing besides that is the following (that could easily be a vector)

I guess I am trying to separate the actual test code which starts on line 214 from the code which is used to initialize the `MerkleStore`. But maybe it is not worth doing it now. Ideally, we'd have a dedicated constructor on the `MerkleStore` - something like:
```Rust
pub fn with_tiered_smt(entries: Vec<(Word, Word)>) -> Self
```
And then we'd just provide a list of key/value pairs to it and computing of node indexes, remaining keys etc. would be handled inside this method.

If we had such a method, this test would look much simpler and I was trying to get closer to this final state. But maybe the best thing is to leave a TODO and come back once we have this method implemented on the `MerkleStore`.

Reply:
Making it panic makes more sense as it better reflects our implementation. Done!

Reply:
Issue added: https://github.com/0xPolygonMiden/crypto/issues/135

Reply:
As side note, the immediately subsequent PR will insert the nodes via `std::collections::smt::insert`, so we will get rid of this problem